openlibrary/core/bookshelves.py

```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
+        for doc in solr_docs:
+            full_key = doc.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "works":
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    docs_dict.update(
+                        {doc["logged_edition"].split("/")[2]: (len(linked_docs) - 1)}
+                    )
+        # Attaches editions to the works, in second loop-- in case of misperformed order.
+        for edition in solr_docs:
+            full_key = edition.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "books":
+                index = docs_dict.get(item_id, -1)
+                if index >= 0:
+                    linked_docs[index].editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning("Error: No work found for edition %s" % full_key)
+        logger.info("LOGGERDICT IS %s" % docs_dict)
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
+        cls,
+        reading_log_work_keys,
+        solr_docs: list[web.Storage],
+        query_params,
+        limit,
+    ) -> list[web.storage]:
+        """
+        Use reading_log_work_keys to fill in missing redirected items in the
+        the solr_docs query results.
+
+        Solr won't return matches for work keys that have been redirected. Because
+        we use Solr to build the lists of storage items that ultimately gets passed
+        to the templates, redirected items returned from the reading log DB will
+        'disappear' when not returned by Solr. This remedies that by filling in
+        dummy works, albeit with the correct work_id.
+        """
+
+        from openlibrary.plugins.worksearch.code import run_solr_query
+        from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
+
+        work_edition_dict = {
+            work_key: edition_key.split("/")[2] if edition_key else None
+            for (work_key, edition_key) in reading_log_work_keys
+        }
+        reading_log_work_keys = [keys[0] for keys in reading_log_work_keys if keys[0]]
+        for idx, work_key in enumerate(reading_log_work_keys):
+            corresponding_solr_doc = next(
+                (doc for doc in solr_docs if doc.key == work_key), None
+            )
+
+            if not corresponding_solr_doc:
+                logger.info("failed key: %s" % work_key)
+                edition_key = work_edition_dict.get(work_key)
+                solr_resp = (
+                    run_solr_query(
+                        scheme=WorkSearchScheme(),
+                        param={'q': '*:*'},
+                        offset=query_params["offset"],
+                        rows=limit,
```

Comment:
We actually don't need these params since we're fetching a specific number of works here instead. This might be the cause of the page 10 error above!

Reply:
Indeed, the source of the error it was! Said error will never rear its ugly head again.

Reply:
Indeed, it was! The dreaded page 10 error has been eliminated, ne'er to rear its ugly head once more. (I hope)